# Project 1 - DataQuest - Introduction to Python

## Analysing Data

As we mentioned in the introduction (showed in notebook ClenData), our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

    Build a minimal Android version of the app, and add it to Google Play.
    If the app has a good response from users, we develop it further.
    If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification. For this analysis, no make sense to built a message exchange app.

Let's begin the analysis by getting a sense of what are the most common genres for each market. For this, we'll need to build frequency tables for a few columns in our data sets.




In [1]:
import csv
from csv import reader
from csv import writer

In [2]:
# AppStore file has Asia char 
import codecs

In [3]:
# TO indentifier Asia char
import re

In [4]:
# To use sort funcion
import operator

In [5]:
# to create a free list
import copy

In [6]:
# to print as a table
from prettytable import PrettyTable

In [7]:
opened_ios = codecs.open(r'C:\Users\Sandra\Documents\Dataquest\AppleStore_Clean.csv','r','utf-8')

In [8]:
opened_gpl = codecs.open(r'C:\Users\Sandra\Documents\Dataquest\googleplaystore_Clean.csv','r','utf-8')

In [9]:
read_ios = reader(opened_ios)
read_gpl = reader(opened_gpl)


In [10]:
data_ios = list(read_ios)
data_gpl = list(read_gpl)

In [11]:
opened_gpl.close()
opened_ios.close()

In [12]:
# print as a table
def print_cabec(row_ios, row_gpl):
    t = PrettyTable(['col', 'IOS', 'GPL'])

    len_ios = len(row_ios)
    len_gpl = len(row_gpl)
    
    app_ios = False
    app_gpl = False

    if len_ios > len_gpl:
        app_ios = True
    else:
        app_gpl = True
    
    if app_ios:  
        for col in range(len(row_gpl)):
            t.add_row([col, row_ios[col], row_gpl[col]])
        for col in range(len(row_gpl), len(row_ios)):
            t.add_row([col, row_ios[col], ''])
        
    if app_gpl:
        for col in range(len(row_ios)):
            t.add_row([col, row_ios[col], row_gpl[col]])
        for col in range(len(row_ios), len(row_gpl)):
            t.add_row([col, '', row_gpl[col], ''])
                       
    print (t)

#t.add_row(['Alice', 24])
#t.add_row(['', 19])
#t.add_row(str)
#print(t)

In [13]:
# print dos cabeçalhos dos arquivos
print_cabec(data_ios[0], data_gpl[0])        

+-----+------------------+----------------+
| col |       IOS        |      GPL       |
+-----+------------------+----------------+
|  0  |                  |      App       |
|  1  |        id        |    Category    |
|  2  |    track_name    |     Rating     |
|  3  |    size_bytes    |    Reviews     |
|  4  |     currency     |      Size      |
|  5  |      price       |    Installs    |
|  6  | rating_count_tot |      Type      |
|  7  | rating_count_ver |     Price      |
|  8  |   user_rating    | Content Rating |
|  9  | user_rating_ver  |     Genres     |
|  10 |       ver        |  Last Updated  |
|  11 |   cont_rating    |  Current Ver   |
|  12 |   prime_genre    |  Android Ver   |
|  13 | sup_devices.num  |                |
|  14 | ipadSc_urls.num  |                |
|  15 |     lang.num     |                |
|  16 |     vpp_lic      |                |
+-----+------------------+----------------+


In [14]:
def print_table(colun_name, colun_qname, titulo, dict_print, percent=False):   
     
    cont_aux = 0
    
    ac_total = 0
    
    for key in dict_print:
        ac_total += float(dict_print[key])
    
    if percent is False:
        t = PrettyTable([colun_name, colun_qname])
        for key in dict_print:
            t.add_row([key[0:60], int(dict_print[key])])
            cont_aux += 1
            if cont_aux > 10:
                break
        
    if percent is True:
        t = PrettyTable([colun_name, colun_qname])
        for key in dict_print:
            t.add_row([key[0:60], round(dict_print[key], 2)])
            cont_aux += 1
            if cont_aux > 10:
                break
    
    if len(dict_print) > 10:
            t.add_row (['...', '...'])
    
    t.add_row(['Total', round(ac_total, 2)])
   
    str_titulo = titulo + ' - Total ' + colun_name + ' ' + str(len(dict_print))
    print (str_titulo, '\n')
    print (t)
    print ('\n')
        
    
    

In [15]:
# contar a qtde de app por genero
# genero 1.9 não diz muito. Vou eliminar esta linha do arquivo
# para IOS - coluna 12
# para GPL - coluna 9

def genre(data_all, col):

    dict_all = {}
    ac_total = 0

    for row in data_all:
        str_genre = row[col]
        if str_genre != '1.9':
            ac_total += 1
            if str_genre in dict_all:
                dict_all[str_genre] += 1
            else:
                dict_all[str_genre] = float(1)

        
    return dict_all, ac_total
        


In [16]:
cabec_ios = 'Dados IOS - Genre'
dict_ios, total_ios = genre(data_ios[1:], 12)

cabec_gpl = "Dados GooglePLay - Category"
dict_gpl_cat, total_gpl_cat = genre(data_gpl[1:], 1)

cabec_gpl = "Dados GooglePLay - Genre"
dict_gpl_gen, total_gpl_gen = genre(data_gpl[1:], 9)


In [17]:
#dict_gpl = sorted(dict_gpl) - este comando transforma o dicionario em uma lista
# sorted(key_value.keys) returns an iterator over the # Dictionary’s value sorted in keys.
# devido a isto foi criado um novo dicionario

def sort_dict(dict_all, skey=True):
    dict_s = {}
    
    if skey is True:
        for i in sorted (dict_all.keys()) : 
             dict_s[i] = dict_all[i] 
    
    if skey is False:
        for i in sorted(dict_all.items(), key= lambda kv: (-kv[1], kv[0])): 
            dict_s[i[0]] = i[1]
                
    return dict_s
    

dict_ios_s = sort_dict(dict_ios, skey=True)
dict_gpl_cat_s = sort_dict(dict_gpl_cat, skey=True)
dict_gpl_gen_s = sort_dict(dict_gpl_gen, skey=True)  
 
print_table('Genre', 'Qty', 'Dados IOS', dict_ios_s, percent=False)
print_table('Genre', 'Qty', 'Dados GPL', dict_gpl_gen_s, percent=False)
print_table('Category', 'Qty', 'Dados GPL', dict_gpl_cat_s, percent=False)

    

Dados IOS - Total Genre 23 

+------------------+--------+
|      Genre       |  Qty   |
+------------------+--------+
|       Book       |   12   |
|     Business     |   17   |
|     Catalogs     |   4    |
|    Education     |  118   |
|  Entertainment   |  252   |
|     Finance      |   35   |
|   Food & Drink   |   26   |
|      Games       |  1861  |
| Health & Fitness |   65   |
|    Lifestyle     |   50   |
|     Medical      |   6    |
|       ...        |  ...   |
|      Total       | 3189.0 |
+------------------+--------+


Dados GPL - Total Genre 115 

+---------------------------------+--------+
|              Genre              |  Qty   |
+---------------------------------+--------+
|              Action             |  275   |
|    Action;Action & Adventure    |   9    |
|            Adventure            |   60   |
|   Adventure;Action & Adventure  |   3    |
|       Adventure;Education       |   1    |
|              Arcade             |  164   |
|    Arcade;Action & Adv

In [18]:
# a qtde de app no GooglePlay difere entre category and gender porque foi retirado 
# a categoria = 1.9 (não significa nada) - mas foi mantida no genero

# classificar por genero ou genero/categoria não nos diz muito.
# Vamos verificar por qtde de app

dict_ios_s = sort_dict(dict_ios, skey=False)
dict_gpl_cat_s = sort_dict(dict_gpl_cat, skey=False)
dict_gpl_gen_s = sort_dict(dict_gpl_gen, skey=False)  
 
print_table('Genre', 'Qty', 'Dados IOS', dict_ios_s, percent=False)
print_table('Genre', 'Qty','Dados GPL', dict_gpl_gen_s, percent=False)
print_table('Category', 'Qty', 'Dados GPL', dict_gpl_cat_s, percent=False)


Dados IOS - Total Genre 23 

+-------------------+--------+
|       Genre       |  Qty   |
+-------------------+--------+
|       Games       |  1861  |
|   Entertainment   |  252   |
|   Photo & Video   |  160   |
|     Education     |  118   |
| Social Networking |  104   |
|      Shopping     |   81   |
|     Utilities     |   77   |
|       Sports      |   69   |
|  Health & Fitness |   65   |
|       Music       |   65   |
|    Productivity   |   55   |
|        ...        |  ...   |
|       Total       | 3189.0 |
+-------------------+--------+


Dados GPL - Total Genre 115 

+-----------------+--------+
|      Genre      |  Qty   |
+-----------------+--------+
|      Tools      |  749   |
|  Entertainment  |  542   |
|    Education    |  476   |
|     Business    |  407   |
|    Lifestyle    |  346   |
|   Productivity  |  346   |
|     Finance     |  328   |
|     Medical     |  313   |
|      Sports     |  307   |
| Personalization |  295   |
|  Communication  |  287   |
|     

In [19]:
# a qtde de app no GooglePlay difere entre category and gender porque foi retirado 
# a categoria = 1.9 (não significa nada) - mas foi mantida no genero

# classificar por genero ou genero/categoria não nos diz muito também. 
# vamos verificar pelo percentual de cada tipo

def perc_app(dict_data, ac_total):
    
    dict_all = {}
    dict_s = {}
    
    for key in dict_data:
        dict_all[key] = (dict_data[key] / ac_total) * 100  
    
    dict_s = sort_dict(dict_all, skey=False)  
            
    return dict_s

dict_ios_p = perc_app(dict_ios, total_ios)
dict_gpl_cat_p = perc_app(dict_gpl_cat, total_gpl_cat)
dict_gpl_gen_p = perc_app(dict_gpl_gen, total_gpl_gen)

print_table('Genre', 'Perc', 'Dados IOS - Frequency Table', dict_ios_p, percent=True)


Dados IOS - Frequency Table - Total Genre 23 

+-------------------+-------+
|       Genre       |  Perc |
+-------------------+-------+
|       Games       | 58.36 |
|   Entertainment   |  7.9  |
|   Photo & Video   |  5.02 |
|     Education     |  3.7  |
| Social Networking |  3.26 |
|      Shopping     |  2.54 |
|     Utilities     |  2.41 |
|       Sports      |  2.16 |
|  Health & Fitness |  2.04 |
|       Music       |  2.04 |
|    Productivity   |  1.72 |
|        ...        |  ...  |
|       Total       | 100.0 |
+-------------------+-------+




We can see that among the free English apps, more than a half (58.36%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.7% of the apps are designed for education, followed by social networking apps which amount for 3.26% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

In [20]:
print_table('Genre', 'Perc', 'Dados GPL - Frequency Table', dict_gpl_gen_p, percent=True)

Dados GPL - Frequency Table - Total Genre 115 

+-----------------+-------+
|      Genre      |  Perc |
+-----------------+-------+
|      Tools      |  8.43 |
|  Entertainment  |  6.1  |
|    Education    |  5.36 |
|     Business    |  4.58 |
|    Lifestyle    |  3.9  |
|   Productivity  |  3.9  |
|     Finance     |  3.69 |
|     Medical     |  3.52 |
|      Sports     |  3.46 |
| Personalization |  3.32 |
|  Communication  |  3.23 |
|       ...       |  ...  |
|      Total      | 100.0 |
+-----------------+-------+





The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:


In [21]:
print_table('Category', 'Perc', 'Dados GPL - Frequency Table', dict_gpl_cat_p, percent=True)

Dados GPL - Frequency Table - Total Category 33 

+-----------------+-------+
|     Category    |  Perc |
+-----------------+-------+
|      FAMILY     | 18.94 |
|       GAME      |  9.7  |
|      TOOLS      |  8.45 |
|     BUSINESS    |  4.58 |
|    LIFESTYLE    |  3.91 |
|   PRODUCTIVITY  |  3.9  |
|     FINANCE     |  3.69 |
|     MEDICAL     |  3.53 |
|      SPORTS     |  3.39 |
| PERSONALIZATION |  3.32 |
|  COMMUNICATION  |  3.23 |
|       ...       |  ...  |
|      Total      | 100.0 |
+-----------------+-------+




The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

### Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre.

For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Below, we calculate the average number of user ratings per app genre on the App Store:


In [22]:
# para elaborar uma tabela de frequencia para installs / rating do app iremos utilizar:
# dict_list = ac_rating, f_rating, qtd_app, dict_app

def frating(data_all, dict_value, col_gen, col_rat, col_app):
    
    dict_rat = {}    

    for row in data_all:
        str_genre = row[col_gen] 
        ac_rating = float(row[col_rat])
        app_name = row[col_app]
        if str_genre in dict_rat:
            dict_rat[str_genre][0] += ac_rating
            dict_rat[str_genre][2] += 1
            dict_rat[str_genre][3][app_name] = float(row[col_rat])
        else:
            dict_rat[str_genre] = [ac_rating, 0, 1, {}]
            dict_rat[str_genre][3][app_name] = float(row[col_rat])
            
    dict_f = {}

    for key in dict_rat:
        freq = dict_rat[key][0] / dict_value[key]
        dict_f[key] = freq
        dict_rat[key][1] = freq
        
    dict_s = sort_dict(dict_f, skey=False) 
        
    return dict_s, dict_rat

In [23]:
cabec_ios = 'Dados IOS - Genre'
dict_ios_frating, dict_ios_rat = frating(data_ios[1:], dict_ios, 12, 6, 2)

In [24]:
def print_table_app(colun_name, colun_qname, titulo, dict_print, dict_s):   
     
    cont_aux = 0
    ac_total = 0
    
    t = PrettyTable([colun_name, colun_qname, 'Qty App'])
            
    for key in dict_s:
        ac_total += 1
        t.add_row([key, str(round(dict_print[key][1], 2)), str(int(dict_print[key][2]))])
        if ac_total > 40:
            t.add_row(['...', '...', '...'])
            break      
   
    str_titulo = titulo + ' - Total ' + colun_name + ' ' + str(len(dict_print))
    print (str_titulo, '\n')
    print (t)
    print ('\n')
    

In [25]:
print_table_app('Genre', 'Freq Rating', 'Dados IOS - Frequency Table', dict_ios_rat, dict_ios_frating)

Dados IOS - Frequency Table - Total Genre 23 

+-------------------+-------------+---------+
|       Genre       | Freq Rating | Qty App |
+-------------------+-------------+---------+
|     Navigation    |   86090.33  |    6    |
|     Reference     |   79350.47  |    17   |
| Social Networking |   72916.55  |   104   |
|       Music       |   58205.03  |    65   |
|      Weather      |   52279.89  |    28   |
|        Book       |   46384.92  |    12   |
|    Food & Drink   |   33333.92  |    26   |
|      Finance      |   32367.03  |    35   |
|       Travel      |   29721.61  |    38   |
|   Photo & Video   |   28441.54  |   160   |
|      Shopping     |   27898.8   |    81   |
|  Health & Fitness |   23298.02  |    65   |
|       Sports      |   23008.9   |    69   |
|       Games       |   22947.76  |   1861  |
|    Productivity   |   21402.8   |    55   |
|        News       |   21248.02  |    43   |
|     Utilities     |   19649.7   |    77   |
|     Lifestyle     |   16815.48 

On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps (see below), which have close to half a million user reviews together. Note that navigation gender has 6 apps in total

In [26]:
# quais os app que estão no genero de Navigation

str_name = 'Navigation'

dict_s = sort_dict(dict_ios_rat[str_name][3], skey=False) 

str_col = str_name + ' - freq rating ' + str(round(dict_ios_rat[str_name][1], 2))
        
print_table(str_col, 'Rating', 'Dados IOS', dict_s, percent=False)


Dados IOS - Total Navigation - freq rating 86090.33 6 

+--------------------------------------------------+----------+
|        Navigation - freq rating 86090.33         |  Rating  |
+--------------------------------------------------+----------+
| Waze - GPS Navigation, Maps & Real-time Traffic  |  345046  |
|        Google Maps - Navigation & Transit        |  154911  |
|                   Geocaching®                    |  12811   |
|   CoPilot GPS – Car Navigation & Offline Maps    |   3582   |
| ImmobilienScout24: Real Estate Search in Germany |   187    |
|               Railway Route Search               |    5     |
|                      Total                       | 516542.0 |
+--------------------------------------------------+----------+




In [27]:
# quais os app que estão no genero de Reference

str_name = 'Reference'

dict_s = sort_dict(dict_ios_rat[str_name][3], skey=False) 

str_col = str_name + ' - freq rating ' + str(round(dict_ios_rat[str_name][1], 2))
        
print_table(str_col, 'Rating', 'Dados IOS', dict_s, percent=False)


Dados IOS - Total Reference - freq rating 79350.47 17 

+--------------------------------------------------------------+-----------+
|               Reference - freq rating 79350.47               |   Rating  |
+--------------------------------------------------------------+-----------+
|                            Bible                             |   985920  |
|            Dictionary.com Dictionary & Thesaurus             |   200047  |
|        Dictionary.com Dictionary & Thesaurus for iPad        |   54175   |
|                       Google Translate                       |   26786   |
|      Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran      |   18418   |
| New Furniture Mods - Pocket Wiki & Game Tools for Minecraft  |   17588   |
|                  Merriam-Webster Dictionary                  |   16849   |
|                          Night Sky                           |   12122   |
| City Maps for Minecraft PE - The Best Maps for Minecraft Poc |    8535   |
| LUCKY BLOCK MOD ™ 

In [28]:
str_name = 'Social Networking'

dict_s = sort_dict(dict_ios_rat[str_name][3], skey=False) 

str_col = str_name + ' - freq rating ' + str(round(dict_ios_rat[str_name][1], 2))
        
print_table(str_col, 'Rating', 'Dados IOS', dict_s, percent=False)


Dados IOS - Total Social Networking - freq rating 72916.55 104 

+--------------------------------------------+-----------+
|  Social Networking - freq rating 72916.55  |   Rating  |
+--------------------------------------------+-----------+
|                  Facebook                  |  2974676  |
|                 Pinterest                  |  1061624  |
|              Skype for iPhone              |   373519  |
|                 Messenger                  |   351466  |
|                   Tumblr                   |   334293  |
|             WhatsApp Messenger             |   287589  |
|                    Kik                     |   260965  |
|  ooVoo – Free Video Call, Text and Voice   |   177501  |
|      TextNow - Unlimited Text + Calls      |   164963  |
|       Viber Messenger – Text & Call        |   164249  |
| Followers - Social Analytics For Instagram |   112778  |
|                    ...                     |    ...    |
|                   Total                    | 758

In [29]:
str_name = 'Book'

dict_s = sort_dict(dict_ios_rat[str_name][3], skey=False) 

str_col = str_name + ' - freq rating ' + str(round(dict_ios_rat[str_name][1], 2))
        
print_table(str_col, 'Rating', 'Dados IOS', dict_s, percent=False)

Dados IOS - Total Book - freq rating 46384.92 12 

+---------------------------------------------------+----------+
|            Book - freq rating 46384.92            |  Rating  |
+---------------------------------------------------+----------+
|    Kindle – Read eBooks, Magazines & Textbooks    |  252076  |
| Audible – audio books, original series & podcasts |  105274  |
|    Color Therapy Adult Coloring Book for Adults   |  84062   |
|     OverDrive – Library eBooks and Audiobooks     |  65450   |
|               HOOKED - Chat Stories               |  47829   |
| BookShout: Read eBooks & Track Your Reading Goals |   879    |
|      Dr. Seuss Treasury — 50 best kids books      |   451    |
|                 Green Riding Hood                 |   392    |
|                  Weirdwood Manor                  |   197    |
|              MangaZERO - comic reader             |    9     |
|           MangaTiara - love comic reader          |    0     |
|                        ...           

However, 'Reference' niche seems to show some potential or may be "Book" niche. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

    Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

    Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

    Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

### Most Popular Apps by Genre on Google Play

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. Let's go to see the app and install for category "Family"

In [30]:
ac_cont = 0
for row in data_gpl:
    if row[1] == 'FAMILY':
        print (row[0], ' ', row[5])
        ac_cont +=1
        if ac_cont > 10:
            break

Jewels Crush- Match 3 Puzzle   1,000,000+
Coloring & Learn   5,000,000+
Mahjong   5,000,000+
Toy Pop Cubes   1,000,000+
Educational Games 4 Kids   5,000,000+
Candy Pop Story   1,000,000+
Candy Smash   1,000,000+
Happy Fruits Bomb - Cube Blast   500,000+
Princess Adventures Puzzles   500,000+
Kids Educational Game 3 Free   5,000,000+
Puzzle Kids - Animals Shapes and Jigsaw Puzzles   1,000,000+


However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.).

One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).


In [31]:
str_value = '5,000,000+'

str_value= str_value.replace(',', '')
str_value= str_value.replace('+', '')
print (str_value)

song = 'cold, cold heart'
replaced_song =  song.replace('o', 'e')

5000000


In [32]:
# para elaborar uma tabela de frequencia para installs iremos utilizar:
# dict_list = ac_rating, f_rating, qtd_app, dict_app

def finstall(data_all, dict_value, col_gen, col_install, col_app):
    
    dict_rat = {}    

    for row in data_all:
        str_genre = row[col_gen] 
        if str_genre != '1.9':
            str_install = row[col_install]
            str_install = str_install.replace(',', '')
            str_install = str_install.replace('+', '')
            ac_install = int(str_install)
            app_name = row[col_app]
            if str_genre in dict_rat:
                dict_rat[str_genre][0] += ac_install
                dict_rat[str_genre][2] += 1
                dict_rat[str_genre][3][app_name] = ac_install
            else:
                dict_rat[str_genre] = [ac_install, 0, 1, {}]
                dict_rat[str_genre][3][app_name] = ac_install
                
            
    dict_f = {}

    for key in dict_rat:
        freq = dict_rat[key][0] / dict_value[key]
        dict_f[key] = freq
        dict_rat[key][1] = freq
        
    dict_s = sort_dict(dict_f, skey=False) 
        
    return dict_s, dict_rat

In [33]:
cabec_gpl = 'Dados GPL - Category'
dict_gpl_frating, dict_gpl_rat = finstall(data_gpl[1:], dict_gpl_cat, 1, 5, 0)

In [34]:
# print (dict_gpl_frating)
print_table_app('Category', 'Freq Install', 'Dados GPL - Frequency Table', dict_gpl_rat, dict_gpl_frating)

Dados GPL - Frequency Table - Total Category 33 

+---------------------+--------------+---------+
|       Category      | Freq Install | Qty App |
+---------------------+--------------+---------+
|    COMMUNICATION    | 38456119.17  |   287   |
|    VIDEO_PLAYERS    | 24727872.45  |   159   |
|        SOCIAL       | 23253652.13  |   236   |
|     PHOTOGRAPHY     | 17772018.76  |   262   |
|     PRODUCTIVITY    | 16738957.55  |   346   |
|         GAME        | 15594505.75  |   861   |
|   TRAVEL_AND_LOCAL  | 13984077.71  |   207   |
|    ENTERTAINMENT    | 11640705.88  |    85   |
|        TOOLS        |  10801391.3  |   750   |
|  NEWS_AND_MAGAZINES |  9401635.95  |   252   |
| BOOKS_AND_REFERENCE |  8631794.09  |   193   |
|       SHOPPING      |  7036877.31  |   199   |
|   PERSONALIZATION   |  5183850.81  |   295   |
|       WEATHER       |  5074486.2   |    71   |
|  HEALTH_AND_FITNESS |  4188821.99  |   273   |
| MAPS_AND_NAVIGATION |  4050083.58  |   123   |
|        FAMILY    

On average, communication apps have the most installs: 38.456.119,17. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [35]:
# para elaborar uma tabela de frequencia para installs iremos utilizar:
# dict_list = ac_rating, f_rating, qtd_app, dict_app

def f_install(data_all, dict_value, col_gen, col_install, col_app, oper):
    
    dict_rat = {}    

    for row in data_all:
        str_genre = row[col_gen] 
        if str_genre != '1.9':
            str_install = row[col_install]
            str_install = str_install.replace(',', '')
            str_install = str_install.replace('+', '')
            ac_install = int(str_install)
            app_name = row[col_app]
            if (oper == '=' and ac_install == 1000000000) or \
            (oper == '>' and ac_install > 1000000000) or \
            (oper == '<' and ac_install < 1000000000):            
                if str_genre in dict_rat:
                    dict_rat[str_genre][0] += ac_install
                    dict_rat[str_genre][2] += 1
                    dict_rat[str_genre][3][app_name] = ac_install
                else:
                    dict_rat[str_genre] = [ac_install, 0, 1, {}]
                    dict_rat[str_genre][3][app_name] = ac_install
                
            
    dict_f = {}

    for key in dict_rat:
        freq = dict_rat[key][0] / dict_value[key]
        dict_f[key] = freq
        dict_rat[key][1] = freq
        
    dict_s = sort_dict(dict_f, skey=False) 
        
    return dict_s, dict_rat

cabec_gpl = 'Dados GPL - Category'
dict_gpl_frating, dict_gpl_rat = f_install(data_gpl[1:], dict_gpl_cat, 1, 5, 0, '=')

str_name = 'COMMUNICATION'

dict_s = sort_dict(dict_gpl_rat[str_name][3], skey=False)    

str_col = str_name + ' - freq install ' + str(round(dict_gpl_rat[str_name][1], 2))
        
print_table(str_col, 'Rating', 'Dados GPL', dict_s, percent=False)

Dados GPL - Total COMMUNICATION - freq install 20905923.34 6 

+------------------------------------------+--------------+
| COMMUNICATION - freq install 20905923.34 |    Rating    |
+------------------------------------------+--------------+
|                  Gmail                   |  1000000000  |
|       Google Chrome: Fast & Secure       |  1000000000  |
|                 Hangouts                 |  1000000000  |
| Messenger – Text and Video Chat for Free |  1000000000  |
|      Skype - free IM & video calls       |  1000000000  |
|            WhatsApp Messenger            |  1000000000  |
|                  Total                   | 6000000000.0 |
+------------------------------------------+--------------+




In [36]:
cabec_gpl = 'Dados GPL - Category'
dict_gpl_frating, dict_gpl_rat = f_install(data_gpl[1:], dict_gpl_cat, 1, 5, 0, '<')

# print (dict_gpl_frating)
print_table_app('Category', 'Freq Install', 'Dados GPL - Frequency Table', dict_gpl_rat, dict_gpl_frating)

str_name = 'COMMUNICATION'

dict_s = sort_dict(dict_gpl_rat[str_name][3], skey=False)    

str_col = str_name + ' - freq install ' + str(round(dict_gpl_rat[str_name][1], 2))
        
print_table(str_col, 'Rating', 'Dados GPL', dict_s, percent=False)



Dados GPL - Frequency Table - Total Category 33 

+---------------------+--------------+---------+
|       Category      | Freq Install | Qty App |
+---------------------+--------------+---------+
|    COMMUNICATION    | 17550195.82  |   281   |
|         GAME        | 14433065.56  |   860   |
|     PHOTOGRAPHY     | 13955224.87  |   261   |
|     PRODUCTIVITY    | 13848784.14  |   345   |
|    VIDEO_PLAYERS    |  12149256.1  |   157   |
|    ENTERTAINMENT    | 11640705.88  |    85   |
|        SOCIAL       | 10541787.72  |   233   |
|        TOOLS        |  9468057.97  |   749   |
|       SHOPPING      |  7036877.31  |   199   |
|  NEWS_AND_MAGAZINES |  5433381.98  |   251   |
|   PERSONALIZATION   |  5183850.81  |   295   |
|       WEATHER       |  5074486.2   |    71   |
|   TRAVEL_AND_LOCAL  |  4322241.96  |   205   |
|  HEALTH_AND_FITNESS |  4188821.99  |   273   |
| MAPS_AND_NAVIGATION |  4050083.58  |   123   |
|        SPORTS       |  3638640.14  |   301   |
| BOOKS_AND_REFEREN

In [37]:
# came back to original values
cabec_gpl = 'Dados GPL - Category'
dict_gpl_frating, dict_gpl_rat = finstall(data_gpl[1:], dict_gpl_cat, 1, 5, 0)

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:


In [38]:
str_name = 'BOOKS_AND_REFERENCE'

dict_s = sort_dict(dict_gpl_rat[str_name][3], skey=False)    

str_col = str_name + ' - freq install ' + str(round(dict_gpl_rat[str_name][1], 2))
        
print_table(str_col, 'Rating', 'Dados GPL', dict_s, percent=False)



Dados GPL - Total BOOKS_AND_REFERENCE - freq install 8631794.09 193 

+-----------------------------------------------+--------------+
| BOOKS_AND_REFERENCE - freq install 8631794.09 |    Rating    |
+-----------------------------------------------+--------------+
|               Google Play Books               |  1000000000  |
|                 Amazon Kindle                 |  100000000   |
|            Audiobooks from Audible            |  100000000   |
|                     Bible                     |  100000000   |
|              Wattpad 📖 Free Books             |  100000000   |
|               Al Quran Indonesia              |   10000000   |
|           Al'Quran Bahasa Indonesia           |   10000000   |
|                Al-Quran (Free)                |   10000000   |
|               Aldiko Book Reader              |   10000000   |
|                  Cool Reader                  |   10000000   |
|                   Dictionary                  |   10000000   |
|                   

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [39]:
str_name = 'BOOKS_AND_REFERENCE'

dict_s = sort_dict(dict_gpl_rat[str_name][3], skey=False)    
for key in dict_s:
    if dict_s[key] == int(1000000) or \
        dict_s[key] == int(5000000) or \
        dict_s[key] == int(10000000) or \
        dict_s[key] == int(50000000):
        print(key, ' ', dict_s[key])
    

Al Quran Indonesia   10000000
Al'Quran Bahasa Indonesia   10000000
Al-Quran (Free)   10000000
Aldiko Book Reader   10000000
Cool Reader   10000000
Dictionary   10000000
Dictionary - Merriam-Webster   10000000
Dictionary.com: Find Definitions for English Words   10000000
English Dictionary - Offline   10000000
English Hindi Dictionary   10000000
FBReader: Favorite Book Reader   10000000
HTC Help   10000000
JW Library   10000000
Moon+ Reader   10000000
NOOK: Read eBooks & Magazines   10000000
Oxford Dictionary of English : Free   10000000
Quran for Android   10000000
Spanish English Translator   10000000
Wikipedia   10000000
50000 Free eBooks & Free AudioBooks   5000000
Al Quran : EAlim - Translations & MP3 Offline   5000000
AlReader -any text book reader   5000000
Ancestry   5000000
Bible KJV   5000000
Dictionary - WordWeb   5000000
Ebook Reader   5000000
English to Hindi Dictionary   5000000
Read books online   5000000
Al Quran Al karim   1000000
All Language Translator Free   1000000


This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

# Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.